### ISSUE TITLE
Elastic search has irregular issues with global search

### ISSUE THREAD

**Issue by vohtaski** [koemei/koemei-widget#246]
related to #240

**Comment by vohtaski** [comment]
Resolution:
after many hours spent trying to locate and understand the problem. I guess now the answer is clear.

Guess 1:
As we could see in #240, there is an issue with connection. I reimplemented the simplest urllib2 instead of urllib3 with connection pool manager inside koemei to call the apis and I still had timeout issue after elastic server restart. Then I started querying with `curl`.

Guess 2:
I thought some settings on production machine were wrong.
I noticed that with simplest curl request, first call was taking 70seconds on production machine after ES restart (that's why connection was lost in previous guess). On my virtual machine it was not the case.
What I found:
1. ES was using 800mb of memory before the request and 2700 after
2. Only one specific search call had such problem (media item with `has_child` transcript)

Guess 3:
Some data needed for this request was not in the memory and it took some time to put it there.
media index has 100000 items
transcript index has 13.5M items
`has_child` query loads these things into memory and it takes time.

Solution:
ElasticSearch has [warmers api](http://www.elasticsearch.org/guide/en/elasticsearch/reference/0.90/indices-warmers.html) that allows to warm up the search query. So the solution is to register the `has_child` query into warmers set of ES and ES should do the rest.

It did solve the problem with elasticsearch restart. Let's hope global search will be always working properly now.

**Comment by vohtaski** [comment]
seems as it only solved the issue on restart, the indexes are cleanup after some period of inactivity :(
will investigate further. options:
- to query the elastic search regularly via cron
- change the query and not use `has_child`

**Comment by vohtaski** [comment]
I think now it's indeed resolved.

The problem was with memory. ES-0.90.0 was loading the child indexes into memory and when garbage collector was running, it did not have enough place in java heap so it was broken.

I upgraded the machine to ES-0.90.13 (latest version of 0.90.x series). I also had to change python APIs because terms highlighting was broken by this update.

0.90.0 was using 1.6Gb of 1.9Gb of java heap, and sometimes after GC+warmer it had `outOfMemory` exception. Now, 0.90.13 goes max to 500MB of 1.9Gb.

I never encountered a problem since then.
